File: dotnet/ThirtyFour.UI.Tests/Util/UtilsTests.cs
using System;
using System.Threading;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using ThirtyFour.UI.Util;

namespace ThirtyFour.UI.Tests.Util
{
    [TestClass]
    public class UtilsTests : BaseTestSuite
    {
        [TestMethod]
        public void TestTimeoutReturnsFalse()
        {
            bool result = Utils.RetryUntilTimeout(TimeSpan.FromSeconds(1), TimeSpan.FromMilliseconds(500), () =>
            {
                Thread.Sleep(1000);
                return false;
            });

            Assert.IsFalse(result);
        }

        [TestMethod]
        public void TestSuccessReturnsTrue()
        {
            bool result = Utils.RetryUntilTimeout(TimeSpan.FromSeconds(1), TimeSpan.FromMilliseconds(500), () =>
            {
                return true;
            });

            Assert.IsTrue(result);
        }

        [TestMethod]
        public void TestFalseResultNotReturnedBeforeTimeout()
        {
            DateTime start = DateTime.UtcNow;

            bool result = Utils.RetryUntilTimeout(TimeSpan.FromMilliseconds(500), TimeSpan.FromMilliseconds(100), () =>
            {
                Thread.Sleep(50);
                return false;
            });

            DateTime end = DateTime.UtcNow;
            TimeSpan diff = end.Subtract(start);

            Assert.IsTrue(diff.Milliseconds > 500);
            Assert.IsFalse(result);
        }

        [TestMethod]
        public void TestSuccessResultReturnedBeforeTimeout()
        {
            DateTime start = DateTime.UtcNow;

            bool result = Utils.RetryUntilTimeout(TimeSpan.FromSeconds(1), TimeSpan.FromMilliseconds(100), () =>
            {
                return true;
            });

            DateTime end = DateTime.UtcNow;
            TimeSpan diff = end.Subtract(start);

            Assert.IsTrue(diff.Milliseconds < 200);
            Assert.IsTrue(result);
        }
    }
}

File: java/src/test/java/com/github/danhumphrey/thirtyfour/ui/util/UtilsTests.java
package com.github.danhumphrey.thirtyfour.ui.util;

import java.util.Date;
import java.util.concurrent.Callable;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import com.github.danhumphrey.thirtyfour.ui.test.BaseTestSuite;

public class UtilsTests extends BaseTestSuite {

	public class TimeoutTask implements Callable<Boolean> {
	    int number;
	 
	    public Boolean call() {
	    	try {
				Thread.sleep(50);
			} catch (InterruptedException e) {
				
			}
            return false;
	    }
	}
	
	public class SuccessTask implements Callable<Boolean> {
	    int number;
	 
	    public Boolean call() {
	    	return true;
	    }
	}
	
	@Test
    public void testTimeoutReturnsFalse() throws Exception {
		boolean result = Utils.retryUntilTimeout(500, 100, new TimeoutTask());
        Assertions.assertFalse(result);
    }

    @Test
    public void testSuccessReturnsTrue() throws Exception {
    	boolean result = Utils.retryUntilTimeout(500, 100, new SuccessTask());
        Assertions.assertTrue(result);
    }
    
    @Test
    public void testFalseResultNotReturnedBeforeTimeout() throws Exception {
    	long start = new Date().getTime(); 
    	boolean result = Utils.retryUntilTimeout(500, 100, new TimeoutTask());
        long stop = new Date().getTime();
        long diff = stop - start;
        Assertions.assertTrue(diff > 500);
        Assertions.assertFalse(result);
    }
    
    @Test()
    public void testTrueResultReturnedBeforeTimeout() throws Exception {
    	long start = new Date().getTime(); 
    	boolean result = Utils.retryUntilTimeout(1000, 100, new SuccessTask());
        long stop = new Date().getTime();
        long diff = stop - start;
        
        System.out.println(diff);
        Assertions.assertTrue(diff < 200);
        Assertions.assertTrue(result);
    }
}
